lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java

```diff
@@ -164,7 +159,7 @@ private void writeChunk(Page page)
         if (bufferedBytes >= writerOption.getMaxRowGroupSize()) {
             columnWriters.forEach(ColumnWriter::close);
             flush();
-            columnWriters.forEach(ColumnWriter::reset);
+            initColumnWriters();
```

Comment:
This bugfix follows the same strategy as in `parquet-mr` project:

https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordWriter.java#L157-L159


After the row group has been flushed, the column writers get reinitialized from scratch.

The `FallbackValuesWriter` doesn't reset both of its writers within the `#reset()` method and this behavior leads to the unwanted behavior documented in https://github.com/trinodb/trino/issues/5518#issuecomment-920304067

Reply:
Comparing https://github.com/trinodb/trino/runs/4650149952?check_suite_focus=true (a test run of this branch) and 
https://github.com/trinodb/trino/runs/4649720577?check_suite_focus=true (a build on `master`) i see differences of about 20% in favor of the `master` implementation compared to the current bugfix implementation  (for the`trino-hive`, `trino-parquet` tests).

Reply:
> The FallbackValuesWriter doesn't reset both of its writers within the #reset() method and this behavior leads to the unwanted behavior documented in #5518 (comment)

When does the FallbackValuesWriter get used?

> One concern which I have with this approach is that the reinitialisation of the writers may affect the performance. I'd need in this regard feedback.

That should not be much of a problem. It happens once per row group and it doesn't seem to be a very expensive operation.

Reply:
> When does the FallbackValuesWriter get used?

It gets used as a fallback for the situation where the dictionary encoding grows too big.

https://github.com/apache/parquet-format/blob/master/Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8